### DISCUSSION TITLE
Roadmap to 1.0

### DISCUSSION THREAD

**Discussion by allison-casey** [hylang/hy#2072] [General]
Hey everybody! I wanted to give y'all an update on what's going on with Hy as we push
towards getting Hy 1.0 out the door and the reasoning behind some of the chnages
we're making in the process.

Hy has been through a pretty extended development process spanning almost seven years
with contributions from many different people at different times. And if you've been
around for some of that, you might know that Hy has been a lot of things throughout
those years (remember when `null` was a thing?).

While contributors, syntax, and language features have changed, the "lipstick on python"
philosophy has stayed the same, and only continued to grown in importance over the years.
Hy 1.0 represents the full embrace of that philosophy. We believe that Hy should be
lispy python. That Hy should embrace all of the things that make Python great and
improve where we can by pulling from lisp tradition. So what's actually changing and why?

1. A general shift towards making things act more like their python counterparts. Making
   `cut` consistent with python `slice`, lambda lists use python's `**`/`*` unpack syntax
   for args and kwargs, not mangling leading `--` for dunder variables, etc
   - **Why?**:: These differences, both in syntax and output, have led to numerous subtle
                bugs and confusion among users when their lispy python doesn't end up
                acting like python.
2. Remove the builtins injection machinery and move the current core hy library of
   utility functions and macros to a dedicated companion python package.
   - **Why?**:: Trying to create a Hy builtins on top of Python's builtins has created
                a number of long standing bugs and unexpected behaviors. The solutions
                that have been proposed to fix this *and* maintain an integrated Hy prelude
                would end up being unacceptably complex or interfere with Hy's ability
                to interop cleanly with Python. In addition, it makes the compiler
                more manageable for us by insulating it from the bikeshedding that
                necessarily occurs around these features which, ultimately, aren't
                core features of the language.

We're a small team, and we want to see Hy reach 1.0 as much as you do. We think that this
is the fastest, cleanest way to make Hy stable and accessible to users in a reasonable
timeframe. You can track the progress we've been making in our 1.0 project board
[here](https://github.com/hylang/hy/projects/1) and in particular the great core
shakeup issue #1992. As always, we're open to suggestions and feedback regarding these
changes, and we hope this answers your questions about where this is all going and why.

**Comment by andrewgiessel** [comment]
Thank you so much for writing this up!  I really appreciate it.  Do we have an estimate or guess as to the 1.0 timeline?  End of summer?  End of year?

**Reply by andrewgiessel** [reply]
non-binding of course =)

**Reply by allison-casey** [reply]
officially, when its done.

In my *non-binding* personal opinion, i would *non-bindingly* be surprised if there wasn't a *hypothetical* something out this year. If that doesn't happen feel free to excoriate me (and only me) from the rooftop of your nearest skyscraper :+1:

**Comment by fergusq** [comment]
Any news on the 1.0 project? The board and the issue linked in the post are closed. Is there a new roadmap?

**Comment by Kodiologist** [comment]
Currently, no. This was Allie's interest, and she hasn't been active on Hy for a few months. My philosophy is not to promise 1.0 until I actually know when it's coming out, and my plan is to wait until we've had a few successive releases with no breaking changes and then call the next one 1.0.

**Comment by Kodiologist** [comment]
Here's an update. I have a vision for Hy 1.0 at this point. It's mostly a matter of attending to a lot of issues that have been open for years. I still don't have a feel for how long it will all take, but it seems Hy time for me to get serious about this.

**Comment by Kodiologist** [comment]
Update: I no longer have any breaking changes planned for Hy 1.0. Best case is that all my further development on Hy before 1.0 is writing documentation and a few bug fixes. Worst case is that some of the latest features I've added turn out to have fundamental problems and I need to rethink things (read: make yet more breaking changes).

**Comment by Kodiologist** [comment]
Hy 1.0.0 is coming out on 2024-09-22 (#2604).
